flink-ml/flink-ml-api/src/main/java/org/apache/flink/ml/api/core/Pipeline.java

```diff
@@ -0,0 +1,217 @@
+	}
+
+	/**
+	 * Appends a PipelineStage to the tail of this pipeline. Pipeline is editable only via this
+	 * method. The PipelineStage must be Estimator, Transformer, Model or Pipeline.
+	 *
+	 * @param stage the stage to be appended
+	 */
+	public Pipeline appendStage(PipelineStage stage) {
+		if (isStageNeedFit(stage)) {
+			lastEstimatorIndex = stages.size();
+		} else if (!(stage instanceof Transformer)) {
+			throw new RuntimeException(
+				"All PipelineStages should be Estimator or Transformer, got:" +
+					stage.getClass().getSimpleName());
+		}
+		stages.add(stage);
+		return this;
+	}
+
+	/**
+	 * Returns a list of all stages in this pipeline in order, the list is immutable.
+	 *
+	 * @return an immutable list of all stages in this pipeline in order.
+	 */
+	public List<PipelineStage> getStages() {
+		return Collections.unmodifiableList(stages);
+	}
+
+	/**
+	 * Check whether the pipeline acts as an {@link Estimator} or not. When the return value is
+	 * true, that means this pipeline contains an {@link Estimator} and thus users must invoke
+	 * {@link #fit(Table)} before they can use this pipeline as a {@link Transformer}. Otherwise,
+	 * the pipeline can be used as a {@link Transformer} directly.
+	 *
+	 * @return {@code true} if this pipeline has an Estimator, {@code false} otherwise
+	 */
+	public boolean needFit() {
+		return this.getIndexOfLastEstimator() >= 0;
+	}
+
+	public Params getParams() {
+		return params;
+	}
+
+	//find the last Estimator or Pipeline that needs fit in stages, -1 stand for no Estimator in Pipeline
+	private int getIndexOfLastEstimator() {
+		return lastEstimatorIndex;
+	}
+
+	@Override
+	/**
+	 * Train the pipeline to fit on the records in the given {@link Table}.
+	 *
+	 * <p>This method go through all the {@link PipelineStage}s in order and does the following
+	 * on each stage until the last {@link Estimator}(inclusive).
+	 *
+	 * <ul>
+	 *     <li>
+	 *         If a stage is an {@link Estimator}, invoke {@link Estimator#fit(Table)} with the input table
+	 *         to generate a {@link Model}, transform the the input table with the generated {@link Model}
+	 *         to get a result table, then pass the result table to the next stage as input.
+	 *     </li>
+	 *     <li>
+	 *         If a stage is a {@link Transformer}, invoke {@link Transformer#transform(Table)} on the input
+	 *         table to get a result table, and pass the result table to the next stage as input.
+	 *     </li>
+	 * </ul>
+	 *
+	 * After all the {@link Estimator}s are trained to fit their input tables, a new pipeline will be created
+	 * with the same stages in this pipeline, except that all the Estimators in the new pipeline are replaced
+	 * with their corresponding Models generated in the above process.
+	 * <p>If there is no {@link Estimator} in the pipeline, the method returns a copy of this pipeline.
+	 */
+	public Pipeline fit(Table input) {
+		List<PipelineStage> transformStages = new ArrayList<>();
```

Comment:
I suggest to set `stages.size();` as initial capacity for `transformStages` list